l2geth/rollup/fees/rollup_fee.go

```diff
@@ -0,0 +1,108 @@
+package fees
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+// overhead represents the fixed cost of batch submission of a single
+// transaction in gas
+const overhead uint64 = 4200
+
+// hundredMillion is a constant used in the gas encoding formula
+const hundredMillion uint64 = 100_000_000
+const hundredBillion uint64 = 100_000_000_000
+const FeeScalar uint64 = 1000
+
+var bigHundredMillion = new(big.Int).SetUint64(hundredMillion)
+var bigHundredBillion = new(big.Int).SetUint64(hundredBillion)
+var BigFeeScalar = new(big.Int).SetUint64(FeeScalar)
+
+// errInvalidGasPrice is the error returned when a user submits an incorrect gas
+// price. The gas price must satisfy a particular equation depending on if it
+// is a L1 gas price or a L2 gas price
+var errInvalidGasPrice = errors.New("rollup fee: invalid gas price")
+
+// CalculateFee calculates the fee that must be paid to the Rollup sequencer, taking into
+// account both the cost of submitting the transaction to L1 as well as
+// executing the transaction on L2
+// fee = (floor((l1GasLimit*l1GasPrice + l2GasLimit*l2GasPrice) / max(tx.gasPrice, 1)) + l2GasLimit) * tx.gasPrice
+// where tx.gasPrice is hard coded to 1000 * wei
+// This function computes tx.gasLimit
+func CalculateRollupFee(data []byte, l1GasPrice, l2GasLimit, l2GasPrice *big.Int) (*big.Int, error) {
+	if err := VerifyGasPrice(l1GasPrice); err != nil {
+		return nil, fmt.Errorf("invalid L1 gas price %d: %w", l1GasPrice, err)
+	}
+	if err := VerifyGasPrice(l2GasPrice); err != nil {
+		return nil, fmt.Errorf("invalid L2 gas price %d: %w", l2GasPrice, err)
+	}
+	l1GasLimit := calculateL1GasLimit(data, overhead)
+	l1Fee := new(big.Int).Mul(l1GasPrice, l1GasLimit)
+	l2Fee := new(big.Int).Mul(l2GasLimit, l2GasPrice)
+	sum := new(big.Int).Add(l1Fee, l2Fee)
+	scaled := new(big.Int).Div(sum, BigFeeScalar)
+	result := new(big.Int).Add(scaled, l2GasLimit)
```

Comment:
Is this correct? You're dividing (i.e. shifting right), shouldn't you be multiplying (i.e. shift left)?

Reply:
This implements the fee equation in the PR notes which does a division. Shifting left would scale the number to be larger and the point is to try to scale the number smaller so that it doesn't look absurdly large. Having it be so large overflows the metamask UI. Open to an alternative proposal where there is a shift left assuming that its cancelled out later